src/workload.rs

```diff
@@ -512,59 +620,173 @@ impl WorkloadInformation {
+                gatewayaddress::Destination::Address(ip) => ip,
+                gatewayaddress::Destination::Hostname(_) => {
+                    info!(%wl.name, "waypoint hostname lookup not supported yet");
+                    return Ok(None); // TODO: implement unsupported feature
+                }
+            };
+            let wp_socket_addr = SocketAddr::new(wp_nw_addr.address, gw_address.port);
+            match self
+                .find_upstream(&wp_nw_addr.network, wp_socket_addr, gw_address.port)
+                .await
+            {
+                Some(upstream) => {
+                    debug!(%wl.name, "found waypoint upstream");
+                    return Ok(Some(upstream));
+                }
+                None => {
+                    debug!(%wl.name, "waypoint upstream not found");
+                    return Err(WaypointError::FindWaypointError(wl.name));
+                }
+            };
+        }
+        Ok(None)
     }
 
     // Support workload and VIP
-    // It is to do on demand workload fetch if necessary, it handles both workload ip and clusterIP
-    async fn fetch_address(&self, addr: &SocketAddr) {
+    // It is to do on demand workload fetch if necessary, it handles both workload ip and services
+    pub async fn fetch_address(&self, network_addr: NetworkAddress) -> Option<Address> {
         // Wait for it on-demand, *if* needed
-        debug!(%addr, "fetch address");
-        // 1. handle workload ip, if workload not found fallback to clusterIP.
-        if self.find_workload(&addr.ip()).is_none() {
-            // 2. handle clusterIP
-            if self.workload_by_vip_exist(addr) {
-                return;
-            }
-            // if both cache not found, start on demand fetch
-            self.fetch_on_demand(&addr.ip()).await;
+        debug!(%network_addr.address, "fetch address");
+        if let Some(address) = self.find_address(network_addr.clone()) {
+            return Some(address);
         }
+        // if both cache not found, start on demand fetch
+        self.fetch_on_demand(&network_addr).await;
+        self.find_address(network_addr)
     }
 
-    async fn fetch_on_demand(&self, ip: &IpAddr) {
+    async fn fetch_on_demand(&self, key: &NetworkAddress) {
         if let Some(demand) = &self.demand {
-            debug!(%ip, "sending demand request");
-            demand.demand(ip.to_string()).await.recv().await;
-            debug!(%ip, "on demand ready");
+            debug!(%key, "sending demand request");
+            demand.demand(key.to_string()).await.recv().await;
+            debug!(%key, "on demand ready");
+        }
+    }
+
+    // keep private so that we can ensure that we always use fetch_address
+    fn find_address(&self, network_addr: NetworkAddress) -> Option<Address> {
+        // 1. handle workload ip, if workload not found fallback to service.
+        match self.find_workload(&network_addr) {
```

Comment:
nit: don't use self.find_workload here. Th at means we lock,unlock 2x. We can just do the locking ourselves so we don't need 2x locking

Reply:
addressed in https://github.com/istio/ztunnel/pull/486/commits/f22b23d2a70ec5cb89d78ccb81b3e4ff1cb2615f